lib/phoenix/router/route.ex

```diff
@@ -96,16 +108,59 @@ defmodule Phoenix.Router.Route do
     end
   end
 
+  defp build_pipes(%Route{kind: :forward} = route) do
+    {_params, fwd_segments} = Plug.Router.Utils.build_path_match(route.path)
+
+    quote do
+      var!(conn)
+      |> Plug.Conn.put_private(:phoenix_pipelines, unquote(route.pipe_through))
+      |> Plug.Conn.put_private(:phoenix_route, fn conn ->
+        opts = unquote(route.plug).init(unquote(route.opts))
+        Phoenix.Router.Route.forward(conn, unquote(fwd_segments), unquote(route.plug), opts)
+      end)
+    end |> pipe_through(route)
+  end
   defp build_pipes(route) do
-    initial = quote do
+    quote do
       var!(conn)
       |> Plug.Conn.put_private(:phoenix_pipelines, unquote(route.pipe_through))
       |> Plug.Conn.put_private(:phoenix_route, fn conn ->
-           opts = unquote(route.controller).init(unquote(route.action))
-           unquote(route.controller).call(conn, opts)
-         end)
-    end
-
+        opts = unquote(route.plug).init(unquote(route.opts))
+        unquote(route.plug).call(conn, opts)
+      end)
+    end |> pipe_through(route)
+  end
+  defp pipe_through(initial, route) do
     Enum.reduce(route.pipe_through, initial, &{&1, [], [&2, []]})
   end
+
+  @doc """
+  Forwards requests to another Plug at a new path.
+  """
+  def forward(%Plug.Conn{path_info: path, script_name: script} = conn, fwd_segments, target, opts) do
+    new_path = path -- fwd_segments
+    {base, ^new_path} = Enum.split(path, length(path) - length(new_path))
+    conn = %{conn | path_info: new_path, script_name: script ++ base} |> target.call(opts)
+    %{conn | path_info: path, script_name: script}
+  end
+
+  @doc """
+  Validates and returns the list of forward path segments.
+
+  Raises RuntimeError plug is already forwarded or path contains
+  a dynamic segment.
+  """
+  def forward_path_segments(path, plug, phoenix_forwards) do
+    case Plug.Router.Utils.build_path_match(path) do
+      {[], path_segments} ->
+        if phoenix_forwards[plug] do
+          raise "`#{inspect plug}` has already been forwarded to. A module can only be forwarded a single time."
+        end
+        path_segments
+      _ ->
+        raise """
+        Dynamic segment `"#{path}"` not allowed when forwarding. Use a static path instead.
```

Comment:
Btw, remember this pattern introduces an extra new line at the end of the message.